[Codegen][TopK] Faster TopK implementation

LLITCHEV opened the pull request:
This is a hybrid (vector and scalars) implementation of TopK.
The load and compare of elements is done using vectors and the
sorting/writing to result is done using scalar registers. The current
implementation is pure scalar.

This results in significant speedup of the function. The speedup scales
with the reduction of the ratio of output/input sizes.

This implementation seems to be about 10% faster even on smaller data
sets where the output/input ration is 1 (see below that benchmark for
input 1x32->1x32, which is basically using TopK for sorting).

For bigger data sets we observe even bigger speedups:
1. For 2x1024 -> 2x40 the speedup is 9.59x
2. For 1x256128 -> 1x40 the speedup is 6.12x.
    
Benchmark numbers - each benchmark was ran 10 times.
### 2x1024 -> 2x40
**Vector**
BM_custom_call_topk_tuple_16/process_time/real_time_mean        0.075 ms        0.103 ms           10 items_per_second=13.2933k/s
BM_custom_call_topk_tuple_16/process_time/real_time_median      0.075 ms        0.102 ms           10 items_per_second=13.3588k/s

**Scalar**
BM_custom_call_topk_tuple_16/process_time/real_time_mean        0.719 ms         1.32 ms           10 items_per_second=1.39137k/s
BM_custom_call_topk_tuple_16/process_time/real_time_median      0.718 ms         1.32 ms           10 items_per_second=1.39307k/s

### 1x32 -> 1x32
**Vector**
BM_custom_call_topk_tuple_16/process_time/real_time_mean        0.042 ms        0.046 ms           10 items_per_second=23.9746k/s
BM_custom_call_topk_tuple_16/process_time/real_time_median      0.042 ms        0.046 ms           10 items_per_second=23.9371k/s

**Scalar**
BM_custom_call_topk_tuple_16/process_time/real_time_mean        0.051 ms        0.055 ms           10 items_per_second=19.5375k/s
BM_custom_call_topk_tuple_16/process_time/real_time_median      0.051 ms        0.055 ms           10 items_per_second=19.6053k/s

### 1x256128 ->1x40
**Vector**
BM_custom_call_topk_tuple_16/process_time/real_time_mean         3.79 ms         3.67 ms           10 items_per_second=264.076/s
BM_custom_call_topk_tuple_16/process_time/real_time_median       3.78 ms         3.67 ms           10 items_per_second=264.459/s

**Scalar**
BM_custom_call_topk_tuple_16/process_time/real_time_mean         23.3 ms         23.2 ms           10 items_per_second=42.9361/s
BM_custom_call_topk_tuple_16/process_time/real_time_median       23.2 ms         23.1 ms           10 items_per_second=43.0286/s

ScottTodd commented on tests/e2e/linalg_ext_ops/top-k-vector.mlir:
This file is too large for an in-tree test. Can you use smaller dimensions (e.g. 256), or generate the file (e.g. https://github.com/openxla/iree/blob/main/tests/e2e/matmul/generate_e2e_matmul_tests.py, https://github.com/openxla/iree/blob/main/tests/e2e/convolution/generate_e2e_conv2d_tests.py)?

GitHub won't even show the diff, had to checkout the branch to see what was taking up so much space :P

```mlir
func.func @vector_topk_test_4x1024_int() {
  %input_values = util.unfoldable_constant dense<[[1, 2, 3, 7, 1, 2, 3, 4, 1, 2, 3, 4,
    1, 2, 3, 4, 1, 2, 3, 5, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4,
    1, 2, 3, 7, 1, 2, 3, 4, 1, 2, 3, 4,
    1, 2, 3, 4, 1, 2, 3, 5, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4,
    1, 2, 3, 7, 1, 2, 3, 4, 1, 2, 3, 4,

...

func.func @vector_call_topk_1x256128() {
  %input_values = util.unfoldable_constant dense<[[
    1.3, 2.0, 3.0, 4.0,  1.0, 2.0, 3.0, 4.0, 1.0, 2.0, 3.0, 4.0, 1.0, 2.0, 3.0,
    4.0, 1.0, 2.0, 3.0,  4.0, 1.0, 2.0, 3.0, 4.0, 1.0, 2.0, 3.0, 4.0, 1.0, 2.0,
...
```

also looks like some auto formatter (clang-format?) ran on the bottom of the file:
```mlir
  ]]> : tensor<1x256128xf32> %
                   out_values_empty = tensor.empty()
      : tensor<1x40xf32> % out_indices_empty = tensor.empty()
      : tensor<1x40xi32> % neg_inf = arith.constant 0xFF800000
      : f32 % c0 = arith.constant 0 : i32 % out_values =
                       linalg
                           .fill
                           ins(% neg_inf
                               : f32)
                               outs(% out_values_empty
                                    : tensor<1x40xf32>) -> tensor<1x40xf32> %
```

LLITCHEV replied:
Fixed.

hanhanW reviewed (dismissed):
Thanks for pushing on this! I skimmed through the changes and I'd like to make some suggestions. It is not a trivial change, so I need more time to review details. Some high-level comments:

1. I think you don't need to update `Passes.h` and `Passes.td` files right? Can you revert the changes?
2. There are methods directly copied from upstream vectorizer... Can we refactor them  in upstream and reuse them here? I can help take a look at code structure if needed.. The point is that we don't want to copy the same code to downstream projects.
3. Please add lit tests to https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/test/generic_vectorization.mlir
4. I think it is always good to vectorize top_k ops. In this context, can you remove the `iree-llvmcpu-vectorize-topk-ops` and `iree-llvmcpu-vectorize-topk` flags? We also need to structure these flags better. I can advice if we really need the flags.

MaheshRavishankar reviewed (commented):
Why are there so many changes to the `Passes.td` file? Is this some linter (in which case, we can just use that as a way to "format" TD files). If not, then maybe better to undo those changes to avoid conflicts.

MaheshRavishankar reviewed (changes requested):
Thanks @LLITCHEV . I understand this is still a draft but high level comment.

Such massive IR changes here need to have lit tests. e2e changes are good but these need lit tests too to see whats happening here. Also please add some high-level description of the IR changes you are looking to do (like the MLIR before and after) which happens with lit tests too, but make it easier to read.

dcaballe reviewed (commented):
Thanks, Lubo! Great work and impressive performance numbers! I know having this working have been tough. Happy to see this reaching the final stage.

> Generating scf ops during vectorization is a big red flag to me because we never use them in vectorizer.

Generating a loop is fine and expected when we "don't know" how to break the loop-carried dependence happening on that dimension. We have been getting away so far because the only dependences we had to deal with were simple reductions that we know how to break. However, for top-k we would need to teach the vectorizer how to break a "K-wide max-index reduction", which is not trivial and would probably result in a very inefficient code (something like VL times K-wide partial reductions that would need to be reduced at the end into the final K-wide reduction). The complexity is not worth it.

The approach taken here is basically to generate a custom vector kernel, or, as we call it for some other ops (e.g., pack/unpack), apply "direct" vectorization. It's true that the code has some complexity but it's implementing a custom sort algorithm so certain level of complexity is expected.

To make the review easier I would suggest that we split the PR into smaller ones. Perhaps we can remove all the code related to legality/preconditions or transformation corner cases and have only the super basic transformation working under a flag, with the corresponding tests, etc... and then follow up with generalizing the transformation and adding all the precondition/legality logic... That should also help the reviewers incrementally build the expertise needed to review and contribute to the code.

Hopefully that helps!

LLITCHEV committed:
Implementation of Topk SCF lowering.

Signed-off-by: Lubo Litchev <lubol@google.com>

LLITCHEV committed:
More topk scf lowering.

Signed-off-by: Lubo Litchev <lubol@google.com>

LLITCHEV committed:
Full implementation of SCF loweing for TopkOp.

Signed-off-by: Lubo Litchev <lubol@google.com>

LLITCHEV committed:
Fixed some linter errors.

Signed-off-by: Lubo Litchev <lubol@google.com>

LLITCHEV committed:
Fixed a test build spec.

Signed-off-by: Lubo Litchev <lubol@google.com>

LLITCHEV committed:
Fixed CMakeList test build target.

Signed-off-by: Lubo Litchev <lubol@google.com>

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Utils/CPUUtils.cpp:
```diff
@@ -76,4 +77,27 @@ bool isLoopPeelingEnabled(FunctionOpInterface funcOp) {
 
+/// Creates a string attribute containing the name of the attribute that is
+/// used to enable SCF lowerig for Topk operations.
+void setEnableTopkSCFLowerAttrName(IREE::LinalgExt::TopkOp topkOp) {
```
Instead of adding a separate attribute name, `lowering_config` now has a dictionary attribute. You can set that as an entry to the dictionary attribute.

LLITCHEV replied:
I did evaluate this at the time of implementing this. The lowering_config doesn't seem to have a dict_attr. The translation config does, but translation config is for funcOps (at least I didn't see how to use it for any operation - like topk). In conversation you mentioned it would be good to add dict_attr to the lowering_config. I agree with that very much! Can we do this a a separate task ad later update this code to use it? Thanks!!!

hanhanW replied:
Can you follow how we add the peeling attribute? I think you can implement something similar to above methods, and follow what has been done for peeling config. E.g.,

https://github.com/iree-org/iree/blob/6fdc30f1638aa8dc5c4af66b91233086376ec530/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L1088-L1093

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Common/TopkLowering.cpp:
```diff
@@ -0,0 +1,915 @@
+                         shapedTy.getShape().end());
+      if (shapedTy.isDynamicShape(vectorSizes))
+        continue;
+      if (failed(lowerTopkOpPreconditionUsingSCF(topkOp))) {
```
Maybe better to check for all precondition during candidate selection and fold all the logic into this method.

LLITCHEV replied:
Done.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Common/TopkLowering.cpp:
```diff
@@ -0,0 +1,915 @@
+}
+
+/// Vectorize a `topKOp` with (1) static result and input types
+static LogicalResult lowerAsLinalgExtTopkUsingSCF(
```
Instead of return by reference, make this return a `FailureOr<SmallVector<Value>>`

LLITCHEV replied:
Changed! Thanks!

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Common/TopkLowering.cpp:
```diff
@@ -0,0 +1,915 @@
+  auto out1 = topkOp.getResults()[1];
+  auto resShapedTy1 = llvm::cast<ShapedType>(out1.getType());
+
+  scf::ForOp scfInputLoop =
```
Why do you expect the `scf.for` op to exist before hand?

LLITCHEV replied:
@MaheshRavishankar Thanks for the chat! It was useful to async on thinks I thought...

LLITCHEV replied:
@MaheshRavishankar 
>>> Why do you expect the scf.for op to exist before hand?
This is the loop that the tiling produces for the TopK operartion... Typically is one tile for each row (vector) in the 2-D case and one tile for the vector case for the whole input. This loop is used as a CodeGen point to generate the helper flags for the SCF lowering of the topk.
The loop is already there - looping over the input elements.
If the loop is not found, the transformation fails and it is delegated to the current implementation that will signal an error and fail.

MaheshRavishankar reviewed (changes requested):
Reading this code I think the lowering of `linalg_ext.topk` to a function call is very hard to maintain. I acknowledge the effort it took to get this all working, but this is fairly involved and fixed function and using the builder API to do this seems extremely hard (which means it is also hard to debug/maintain).

I think the last time we discussed I had suggested using something more aligned with the way microkernels are lowered. Let me try to spell it out a bit more.

In its current form, I think there is a lot of complexity introduced since this is being done before bufferization. I am assuming that it would be easier if this was done post bufferization (though the check for an scf.for before seems like a problem I am ignoring for now). So for the top-k dispatch, post bufferization you will have something like this

```
module {
  func.func @dispatch() {
    iree_linalg_ext.topk ins(....) outs(....)
  }
}
```

You can then pattern match this and replace it with a 

```
module {
  func.func @topk_implementation(...)
  func.func @dispatch() {
    func.call @topk_implementation(....)
  }
}
```

Then you can populate the body of the `func.func @topk_implementation` appropriately. For now you can just take the implementation you have in `TopKLowering.cpp` and use that to "generate" the body of this function.

A more maintainable way of doing this would be we embed the implementation of `func.func @topk_implementation(...)` in a `*.mlirbc` file that is shipped with IREE and loaded into the HAL as a deserialized object (very much the same way we have `*.bc` files that contain the implementation of `linalg.mmt4d` (and other operations) operations. Then a pass in the codegen backend can "link" and inline the implementation of topk. The advantage is that you wont have to struggle with the builder (and understand the builder code) to see what IR is being generated. You could just read the implementation in MLIR. That seems more maintainable. We can get there in stages though. If for now you just want to move all the builder code you have as part of lowering `linalg_ext.topk` into a method that populates the body of a `func.func`, we can land that and then see how we can just ship an MLIR file directly for the implementation as a follow up.

LLITCHEV committed:
Merge remote-tracking branch 'upstream' into tk-fix

LLITCHEV committed:
Return values as FailureOr<> instead of ref

Signed-off-by: Lubo Litchev <lubol@google.com>

LLITCHEV committed:
Merge branch 'main' of https://github.com/openxla/iree into tk-fix

LLITCHEV committed:
Addressed some CR comments.

Moved the test for input dyn_size to the decision maker for lowering of
topk from the precondition test.

Signed-off-by: Lubo Litchev <lubol@google.com>

LLITCHEV committed:
Fixed some formatting issues.

Signed-off-by: Lubo Litchev <lubol@google.com>

LLITCHEV committed:
Remove some CUDA tests.

The current implementation is producing wrong results.

Signed-off-by: Lubo Litchev <lubol@google.com>

LLITCHEV committed:
Remove the entry not usin gthe exclude directive.

Signed-off-by: Lubo Litchev <lubol@google.com>

MaheshRavishankar reviewed (approved):
Aprooving as an in-tree experiment. This helps in some use cases today, but will need some follow up work to make more maintainable.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2501,6 +2506,27 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
+  unsigned typeWidthInBytes = IREE::Util::getRoundedElementByteWidth(
+      topkOp.getInputType().getElementType());
+  int64_t typeVectorSize = getVectorSize(entryPointFn, typeWidthInBytes);
+  SmallVector<int64_t> vecTileSizes{0, typeVectorSize};
```
Use assignment style to avoid potential bugs and for consistency.

```suggestion
  SmallVector<int64_t> vecTileSizes = {0, typeVectorSize};
```

https://abseil.io/tips/88

btw, don't we need to take numLoops into account?

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2501,6 +2506,27 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
+  SmallVector<int64_t> vecTileSizes{0, typeVectorSize};
+  SmallVector<int64_t> distTileSizes = getDefaultDistributedLevelTileSizes(
+      topkOp, DistributionHeuristicConfig{});
+  SmallVector<int64_t> lbs, ubs;
+  getRangeBounds(cast<TilingInterface>(topkOp.getOperation()), lbs, ubs);
+  int64_t numLoops = lbs.size();
```
IIRC, you only need the number of loops, so the `ubs` is not used? In this context, can we rewrite it to `int64_t numLoops = cast<TilingInterface>(topkOp.getOperation()).getLoopIteratorTypes().size();`?

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2513,6 +2539,54 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
+  if (!clSCFLowerTopk)
+    return false;
+  auto arg0 = topkOp.getInputs()[0];
+  auto shapedTy = llvm::cast<ShapedType>(arg0.getType());
+  SmallVector<int64_t> vectorSizes;
+  vectorSizes.append(shapedTy.getShape().begin(), shapedTy.getShape().end());
```
nit1: use `cast<>` directly for consistency.

nit2: You can collapse `append` with initializer, e.g.,

```suggestion
  auto shapedTy = cast<ShapedType>(arg0.getType());
  SmallVector<int64_t> vectorSizes(shapedTy.getShape());
```

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2513,6 +2539,54 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
+  auto shapedTy = llvm::cast<ShapedType>(arg0.getType());
+  SmallVector<int64_t> vectorSizes;
+  vectorSizes.append(shapedTy.getShape().begin(), shapedTy.getShape().end());
+  if (shapedTy.isDynamicShape(vectorSizes))
+    return false;
```
I think you can use `!shapedTy.hasStaticShape()` here.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2513,6 +2539,54 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
+    return false;
+  auto out0 = topkOp.getResults()[0];
+  auto resShapedTy = llvm::cast<ShapedType>(out0.getType());
+  ArrayRef<int64_t> resShape = resShapedTy.getShape();
+  if (resShapedTy.isDynamicShape(resShape))
+    return false;
```
same here, you can use the above method, then you don't need to the `resShape` variable.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2513,6 +2539,54 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
+  if (resShapedTy.isDynamicShape(resShape))
+    return false;
+
+  auto inShapedTy = topkOp.getInputType();
+  ArrayRef<int64_t> inShape = inShapedTy.getShape();
```
nit: collapse it into a single line because `inShapedTy` is not used in other places.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2513,6 +2539,54 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
+
+  // Validate input
+  auto inElemTy = topkOp.getInputType().getElementType();
+  if (auto intType = llvm::dyn_cast_if_present<IntegerType>(inElemTy)) {
+    // Do nothing. Expected type.
+  } else if (auto floatType = llvm::dyn_cast_if_present<FloatType>(inElemTy)) {
+    // Do nothing. Expected type.
+  } else {
+    // Unexpected type.
+    return false;
+  }
```
I think this can be simplified by using `isIntOrFloat()`. E.g.,

```
if (!inElemTy.isIntOrFloat()) {
  return false;
}
```

https://github.com/llvm/llvm-project/blob/de1f5b96adcea52bf7c9670c46123fe1197050d2/mlir/include/mlir/IR/Types.h#L173

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -2513,6 +2539,54 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
+
+  // Make sure the number of elements of the input's last dim is multiple of
+  // the number of elements in a register.
+  if (vectorSizes[vectorSizes.size() - 1] % typeVectorSize != 0)
```
nit: `vectorSizes.back()`

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:
```diff
@@ -621,6 +626,7 @@ void addCPULinalgExtTileAndVectorizePipeline(
     options.useConfiguredVectorSizes = pipelineOpt.useConfiguredVectorSizes;
     options.enableVectorMasking = pipelineOpt.enableVectorMasking;
     funcPassManager.addPass(createGenericVectorizationPass(options));
+    funcPassManager.addPass(createTopkLoweringPass());
```
You only need `createTopkLoweringPass` in this pipeline. Can you remove other changes?

hanhanW commented on tests/e2e/linalg_ext_ops/BUILD.bazel:
```diff
@@ -135,18 +139,30 @@ iree_check_single_backend_test_suite(
 # TODO: Remove topk-split-reduction tests? The flag looks like an unfinished experiment.
-
+#
+# Disable the top-k-scf-lowering.mlir test.
+# The following error in the split-reduction code.ßßß
+# [ RUN      ] module.vector_call_topk_1x256
+# iree/runtime/src/iree/modules/check/module.cc:375: Failure
+# Failed
+# Expected equality of these values. Contents does not match.
+#   lhs:
+#     1x40xi32=[135 135 169 169 247 247 233 233 3 3 7 7 11 11 15 15 19 19 23 23 27 27 31 31 35 35 39 39 43 43 47 47 51 51 55 55 59 59 63 63]
+#   rhs:
+#    1x40xi32=[135 169 247 233 3 7 11 15 19 23 27 31 35 39 43 47 51 55 59 63 67 71 75 79 83 87 91 95 99 103 107 111 115 119 123 127 131 139 143 147]
```
It is better to create an issue and attach the issue ID here. So people can give updates or discuss things on the issue when needed.

hanhanW commented on tests/e2e/linalg_ext_ops/top-k-scf-lowering.mlir:
```diff
@@ -0,0 +1,193 @@
+  return
+}
+
+// int32 type not supported in the scalar implementation.
```
Can we remove them from the mlir file? Are we going to support that? If so, you can add the TODO to `KernelDispatch.cpp` instead of commenting them out.

hanhanW reviewed (changes requested):
Thanks! Here is the first round of review comments. I haven't reviewed TopkLoweringPass.cpp yet.
